src/pip/_internal/index.py

```diff
@@ -939,15 +939,25 @@ def _get_encoding_from_headers(headers):
     return None
 
 
-_CLEAN_LINK_RE = re.compile(r'[^a-z0-9$&+,/:;=?@.#%_\\|-]', re.I)
-
-
 def _clean_link(url):
     # type: (str) -> str
     """Makes sure a link is fully encoded.  That is, if a ' ' shows up in
     the link, it will be rewritten to %20 (while not over-quoting
     % or other characters)."""
-    return _CLEAN_LINK_RE.sub(lambda match: '%%%2x' % ord(match.group(0)), url)
+    # Split url into parts.
+    result = urllib_parse.urlparse(url)
+    if result.netloc == "":
+        # On Windows the path part might contain a drive letter which
+        # should not be quoted. On Linux there are no drive letters,
+        # and a colon should be quoted. We rely on urllib.request to
+        # do the right thing here.
+        path = urllib_request.pathname2url(
+            urllib_request.url2pathname(result.path))
+    else:
+        # Re-Quote URL and make sure we are not double quoting anything.
```

Comment:
Since this comment applies to both the file URL case and the normal case, I'd explain this comment above the if condition (since in both cases you are using the same trick of unquoting before quoting).

Reply:
Done.